java/query/src/main/java/sleeper/query/model/QuerySerDe.java

```diff
@@ -126,18 +145,35 @@ public Query fromJson(String jsonQuery) {
     }
 
     public static class QueryJsonSerDe implements JsonSerializer<Query>, JsonDeserializer<Query> {
-        private final Function<String, Schema> tableNameToSchemaFunction;
+        private final SchemaLoader schemaLoader;
 
-        public QueryJsonSerDe(Function<String, Schema> tableNameToSchemaFunction) {
-            this.tableNameToSchemaFunction = tableNameToSchemaFunction;
+        public QueryJsonSerDe(SchemaLoader schemaLoader) {
+            this.schemaLoader = schemaLoader;
         }
 
         public QueryJsonSerDe(TablePropertiesProvider tablePropertiesProvider) {
-            this(tableName -> tablePropertiesProvider.getByName(tableName).getSchema());
+            this((queryId, tableName) -> {
+                if (tableName == null) {
+                    throw new QueryValidationException(queryId, "Table must not be null");
+                }
+                try {
+                    return tablePropertiesProvider.getByName(tableName).getSchema();
+                } catch (NoSuchElementException e) {
+                    throw new QueryValidationException(queryId, "Table \"" + tableName + "\" does not exist");
+                }
+            });
         }
 
         public QueryJsonSerDe(Map<String, Schema> tableNameToSchemaMap) {
-            this(tableNameToSchemaMap::get);
+            this((queryId, tableName) -> {
+                if (tableName == null) {
+                    throw new QueryValidationException(queryId, "Table must not be null");
+                }
+                if (!tableNameToSchemaMap.containsKey(tableName)) {
+                    throw new QueryValidationException(queryId, "Table \"" + tableName + "\" does not exist");
+                }
+                return tableNameToSchemaMap.get(tableName);
```

Comment:
This seems to duplicate the behaviour in the constructor that takes TablePropertiesProvider. The null check of tableName doesn't look like it needs to be here. The loading the schema or throwing the exception could use a functional interface that takes the table name and returns Optional<Schema>? That'd be similar to how it was before, and would let us have the validation behaviour outside the lambda.